File: Calc.Tests/CalculatorTests.cs
using System;
using FsCheck;
using FsCheck.Xunit;
using Xunit;
using Xunit.Abstractions;

namespace Calc.Tests
{
    public class CalculatorTests
    {
        private readonly ITestOutputHelper _outputHelper;

        public CalculatorTests(ITestOutputHelper outputHelper)
        {
            _outputHelper = outputHelper;
        }

        [Fact]
        public void Add_AddingTwoPositiveNumbers1And2_ResultIs3()
        {
            var calculator = CreateCalculator();

            calculator.Add(1, 2);

            Assert.Equal(3, calculator.Result);
        }

        [Fact]
        public void Add_AddingTwoNegativeNumbersNegative1AndNegative2_ResultIsNegative3()
        {
            var calculator = CreateCalculator();

            calculator.Add(-1, -2);

            Assert.Equal(-3, calculator.Result);
        }

        [Fact]
        public void Add_AddingNegativeAndPositveNumbers1AndNegative2_ResultIsNegative1()
        {
            var calculator = CreateCalculator();

            calculator.Add(1, -2);

            Assert.Equal(-1, calculator.Result);
        }

        [Theory]
        [InlineData(1, 2, 3)]
        [InlineData(-1, -2, -3)]
        [InlineData(1, -2, -1)]
        [InlineData(100, 200, 300)]
        public void Add_TwoNumbers_ResultMustBeTheExpectedValue(int x, int y, int expected)
        {
            var calculator = CreateCalculator();

            calculator.Add(x, y);

            Assert.Equal(expected, calculator.Result);
        }

        [Property]
        public bool Add_TwoNumbers_Result(int x, int y)
        {
            var calculator = CreateCalculator();

            calculator.Add(x, y);

            _outputHelper.WriteLine($"x: {x}, y: {y}, result: {calculator.Result}");

            return calculator.Result == x + y;
        }

        private static Calculator CreateCalculator() => new Calculator();
    }
}

File: Calc.Fsharp.FsCheck.Tests/CalculatorTests.fs
module ``Calculator Tests``

open Xunit
open Calc
open FsCheck
open FsCheck.Xunit
open FsCheck.Gen

let add x y =
    let calculator = new Calculator()
    calculator.Add(x, y)
    calculator.Result

let equal (expected:int) (actual:int) =
    Assert.Equal(expected, actual)

[<Fact>]
let ``When two positive numbers 1 and 2 added then 3 must be returned``() =

    add 1 2 |> equal 3

[<Fact>]
let ``When two negative numbers -1 and - 2 added then -3 must be returned``() =

    add -1 -2 |> equal -3

[<Fact>]
let ``When 1 and -2 is added then -1 must be returned``() =

    add 1 -2 |> equal -1

[<Theory>]
[<InlineData(1, 2, 3)>]
[<InlineData(-1, -2, -3)>]
[<InlineData(1, -2, -1)>]
[<InlineData(100, 200, 300)>]
let ``When two number x and y added then an expected number must be retuned``(x, y, expectedValue) =

    add x y |> equal expectedValue

[<Fact>]
let ``When two number added then an expected number must be retuned``() =

    let ``Add x, y must be x + y`` x y =
        add x y = x + y

    Check.QuickThrowOnFailure ``Add x, y must be x + y``

[<Property>]
let ``When two number added then an expected number must be retuned (v2)``(x, y) =

    add x y = x + y
